File: hello-websocket-ts/common/codec.ts
// Hello WebSocket Protocol Codec - TypeScript implementation.
// Implements the canonical binary protocol defined in PROTOCOL.md.

import { createHash } from 'crypto';

// ─── Constants ───────────────────────────────────────────────────────────

export const PORT = 9898;
export const MAGIC = 0x48;
export const VERSION = 0x01;
export const HEADER_LEN = 8;
export const SERVER_LANG = 'TYPESCRIPT';
export const CLIENT_LANG = 'TYPESCRIPT';

// Message types
export const MSG_HELLO = 0x01;
export const MSG_BONJOUR = 0x02;
export const MSG_ECHO_REQUEST = 0x03;
export const MSG_ECHO_RESPONSE = 0x04;
export const MSG_KISS_REQUEST = 0x05;
export const MSG_KISS_RESPONSE = 0x06;
export const MSG_PING = 0x07;
export const MSG_PONG = 0x08;
export const MSG_TIME_NOTIFICATION = 0x09;
export const MSG_RANDOM_NUMBER = 0x0A;
export const MSG_HASH_RESPONSE = 0x0B;
export const MSG_DISCONNECT = 0x0C;
export const MSG_ERROR = 0x7F;

// Error codes
export const ERR_DECODE = 0x01;
export const ERR_UNKNOWN_MSG_TYPE = 0x02;
export const ERR_TRUNCATED_PAYLOAD = 0x03;
export const ERR_BAD_MAGIC = 0x04;
export const ERR_BAD_VERSION = 0x05;
export const ERR_SESSION_NOT_FOUND = 0x06;
export const ERR_INTERNAL = 0x07;

// Intervals (ms)
export const PING_INTERVAL_MS = 1000;
export const SESSION_TIMEOUT_MS = 60000;
export const TIME_INTERVAL_MS = 5000;
export const RANDOM_INTERVAL_MS = 5000;
export const KISS_INTERVAL_MS = 5000;

// ─── Types ──────────────────────────────────────────────────────────────

export interface EchoResult {
  idx: bigint;
  type: number;
  kv: Record<string, string>;
}

export type Message =
  | { type: typeof MSG_HELLO; clientLanguage: string }
  | { type: typeof MSG_BONJOUR; serverLanguage: string }
  | { type: typeof MSG_ECHO_REQUEST; echoId: bigint; echoMeta: string; echoData: string }
  | { type: typeof MSG_ECHO_RESPONSE; echoStatus: number; echoResults: EchoResult[] }
  | { type: typeof MSG_KISS_REQUEST; osName: string; osVersion: string; osRelease: string; osArch: string }
  | { type: typeof MSG_KISS_RESPONSE; kissLanguage: string; kissEncoding: string; kissTimeZone: string }
  | { type: typeof MSG_PING; timestampMs: bigint }
  | { type: typeof MSG_PONG; timestampMs: bigint }
  | { type: typeof MSG_TIME_NOTIFICATION; timestampMs: bigint; iso8601: string }
  | { type: typeof MSG_RANDOM_NUMBER; randomId: bigint; randomNumber: bigint }
  | { type: typeof MSG_HASH_RESPONSE; randomId: bigint; hashHex: string }
  | { type: typeof MSG_DISCONNECT; disconnectReason: string }
  | { type: typeof MSG_ERROR; errorCode: number; errorMessage: string };

// ─── ByteWriter ─────────────────────────────────────────────────────────

export class ByteWriter {
  private buf: number[] = [];

  writeU8(v: number): void { this.buf.push(v & 0xFF); }

  writeU16(v: number): void { this.buf.push((v >> 8) & 0xFF, v & 0xFF); }

  writeU32(v: number): void {
    this.buf.push((v >>> 24) & 0xFF, (v >> 16) & 0xFF, (v >> 8) & 0xFF, v & 0xFF);
  }

  writeI32(v: number): void { this.writeU32(v >>> 0); }

  writeI64(v: bigint): void {
    const lo = Number(v & 0xFFFFFFFFn);
    const hi = Number((v >> 32n) & 0xFFFFFFFFn);
    this.writeU32(hi);
    this.writeU32(lo);
  }

  writeString(s: string): void {
    const b = Buffer.from(s, 'utf-8');
    this.writeU32(b.length);
    for (const byte of b) this.buf.push(byte);
  }

  writeKV(m: Record<string, string>): void {
    const keys = Object.keys(m);
    this.writeU32(keys.length);
    for (const k of keys) { this.writeString(k); this.writeString(m[k]); }
  }

  toBuffer(): Buffer { return Buffer.from(this.buf); }
}

// ─── ByteReader ────────────────────────────────────────────────────────

export class ByteReader {
  private data: Buffer;
  private pos: number = 0;

  constructor(data: Buffer) { this.data = data; }

  remaining(): number { return this.data.length - this.pos; }
  private ensure(size: number, field: string): void {
    if (!Number.isSafeInteger(size) || size < 0 || size > this.remaining()) {
      throw new Error(`${field} requires ${size} bytes, only ${this.remaining()} remain`);
    }
  }

  readU8(): number { this.ensure(1, 'u8'); return this.data[this.pos++]; }

  readU16(): number {
    this.ensure(2, 'u16');
    const v = (this.data[this.pos] << 8) | this.data[this.pos + 1];
    this.pos += 2;
    return v;
  }

  readU32(): number {
    this.ensure(4, 'u32');
    const v = (this.data[this.pos] * 0x1000000) + (this.data[this.pos + 1] << 16) +
              (this.data[this.pos + 2] << 8) + this.data[this.pos + 3];
    this.pos += 4;
    return v;
  }

  readI32(): number { return this.readU32() | 0; }

  readI64(): bigint {
    const hi = BigInt(this.readU32());
    const lo = BigInt(this.readU32());
    return BigInt.asIntN(64, (hi << 32n) | lo);
  }

  readString(): string {
    const ln = this.readU32();
    this.ensure(ln, 'string');
    const s = this.data.toString('utf-8', this.pos, this.pos + ln);
    this.pos += ln;
    return s;
  }

  readKV(): Record<string, string> {
    const count = this.readU32();
    if (count > Math.floor(this.remaining() / 8)) throw new Error(`kv count ${count} exceeds remaining payload`);
    const m: Record<string, string> = {};
    for (let i = 0; i < count; i++) {
      const k = this.readString();
      const v = this.readString();
      m[k] = v;
    }
    return m;
  }
}

// ─── Frame Codec ────────────────────────────────────────────────────────

export function encodeFrame(msgType: number, payload: Buffer): Buffer {
  const header = Buffer.alloc(HEADER_LEN);
  header[0] = MAGIC;
  header[1] = VERSION;
  header[2] = msgType;
  header[3] = 0x00;
  header.writeUInt32BE(payload.length, 4);
  return Buffer.concat([header, payload]);
}

export function decodeFrame(data: Buffer): { msgType: number; payload: Buffer } {
  if (data.length < HEADER_LEN) throw new Error(`frame too short: ${data.length}`);
  if (data[0] !== MAGIC) throw new Error(`bad magic: 0x${data[0].toString(16)}`);
  if (data[1] !== VERSION) throw new Error(`bad version: 0x${data[1].toString(16)}`);
  const msgType = data[2];
  const payloadLen = data.readUInt32BE(4);
  if (payloadLen !== data.length - HEADER_LEN) {
    throw new Error(`payload length mismatch: declared ${payloadLen}, available ${data.length - HEADER_LEN}`);
  }
  return { msgType, payload: data.subarray(HEADER_LEN, HEADER_LEN + payloadLen) };
}

// ─── Message Encoders ───────────────────────────────────────────────────

export function encodeMessage(msg: Message): Buffer {
  const w = new ByteWriter();
  switch (msg.type) {
    case MSG_HELLO: w.writeString(msg.clientLanguage); return encodeFrame(MSG_HELLO, w.toBuffer());
    case MSG_BONJOUR: w.writeString(msg.serverLanguage); return encodeFrame(MSG_BONJOUR, w.toBuffer());
    case MSG_ECHO_REQUEST: w.writeI64(msg.echoId); w.writeString(msg.echoMeta); w.writeString(msg.echoData); return encodeFrame(MSG_ECHO_REQUEST, w.toBuffer());
    case MSG_ECHO_RESPONSE:
      w.writeI32(msg.echoStatus); w.writeU32(msg.echoResults.length);
      for (const r of msg.echoResults) { w.writeI64(r.idx); w.writeU8(r.type); w.writeKV(r.kv); }
      return encodeFrame(MSG_ECHO_RESPONSE, w.toBuffer());
    case MSG_KISS_REQUEST: w.writeString(msg.osName); w.writeString(msg.osVersion); w.writeString(msg.osRelease); w.writeString(msg.osArch); return encodeFrame(MSG_KISS_REQUEST, w.toBuffer());
    case MSG_KISS_RESPONSE: w.writeString(msg.kissLanguage); w.writeString(msg.kissEncoding); w.writeString(msg.kissTimeZone); return encodeFrame(MSG_KISS_RESPONSE, w.toBuffer());
    case MSG_PING: w.writeI64(msg.timestampMs); return encodeFrame(MSG_PING, w.toBuffer());
    case MSG_PONG: w.writeI64(msg.timestampMs); return encodeFrame(MSG_PONG, w.toBuffer());
    case MSG_TIME_NOTIFICATION: w.writeI64(msg.timestampMs); w.writeString(msg.iso8601); return encodeFrame(MSG_TIME_NOTIFICATION, w.toBuffer());
    case MSG_RANDOM_NUMBER: w.writeI64(msg.randomId); w.writeI64(msg.randomNumber); return encodeFrame(MSG_RANDOM_NUMBER, w.toBuffer());
    case MSG_HASH_RESPONSE: w.writeI64(msg.randomId); w.writeString(msg.hashHex); return encodeFrame(MSG_HASH_RESPONSE, w.toBuffer());
    case MSG_DISCONNECT: w.writeString(msg.disconnectReason); return encodeFrame(MSG_DISCONNECT, w.toBuffer());
    case MSG_ERROR: w.writeI32(msg.errorCode); w.writeString(msg.errorMessage); return encodeFrame(MSG_ERROR, w.toBuffer());
  }
}

// ─── Message Decoder ────────────────────────────────────────────────────

export function decodeMessage(data: Buffer): Message {
  const { msgType, payload } = decodeFrame(data);
  const r = new ByteReader(payload);
  switch (msgType) {
    case MSG_HELLO: return { type: MSG_HELLO, clientLanguage: r.readString() };
    case MSG_BONJOUR: return { type: MSG_BONJOUR, serverLanguage: r.readString() };
    case MSG_ECHO_REQUEST: return { type: MSG_ECHO_REQUEST, echoId: r.readI64(), echoMeta: r.readString(), echoData: r.readString() };
    case MSG_ECHO_RESPONSE: {
      const status = r.readI32();
      const count = r.readU32();
      if (count > Math.floor(r.remaining() / 13)) throw new Error(`result count ${count} exceeds remaining payload`);
      const results: EchoResult[] = [];
      for (let i = 0; i < count; i++) results.push({ idx: r.readI64(), type: r.readU8(), kv: r.readKV() });
      return { type: MSG_ECHO_RESPONSE, echoStatus: status, echoResults: results };
    }
    case MSG_KISS_REQUEST: return { type: MSG_KISS_REQUEST, osName: r.readString(), osVersion: r.readString(), osRelease: r.readString(), osArch: r.readString() };
    case MSG_KISS_RESPONSE: return { type: MSG_KISS_RESPONSE, kissLanguage: r.readString(), kissEncoding: r.readString(), kissTimeZone: r.readString() };
    case MSG_PING: return { type: MSG_PING, timestampMs: r.readI64() };
    case MSG_PONG: return { type: MSG_PONG, timestampMs: r.readI64() };
    case MSG_TIME_NOTIFICATION: return { type: MSG_TIME_NOTIFICATION, timestampMs: r.readI64(), iso8601: r.readString() };
    case MSG_RANDOM_NUMBER: return { type: MSG_RANDOM_NUMBER, randomId: r.readI64(), randomNumber: r.readI64() };
    case MSG_HASH_RESPONSE: return { type: MSG_HASH_RESPONSE, randomId: r.readI64(), hashHex: r.readString() };
    case MSG_DISCONNECT: return { type: MSG_DISCONNECT, disconnectReason: r.readString() };
    case MSG_ERROR: return { type: MSG_ERROR, errorCode: r.readI32(), errorMessage: r.readString() };
    default: throw new Error(`unknown message type: 0x${msgType.toString(16)}`);
  }
}

// ─── Utility ────────────────────────────────────────────────────────────

export function nowMs(): bigint { return BigInt(Date.now()); }

export function nowISO(): string { return new Date().toISOString().replace(/\.\d+Z$/, 'Z'); }

export function hashNumber(num: bigint): string {
  const h = createHash('sha256').update(num.toString()).digest('hex');
  return h.substring(0, 10);
}

export function log(name: string, msg: string): void {
  const ts = new Date().toISOString().replace('T', ' ').substring(0, 19);
  console.log(`[${ts}] [INFO] [${name}] ${msg}`);
}

// ─── Message Factory Helpers ─────────────────────────────────────────────

export const hello = (clientLanguage: string): Message => ({ type: MSG_HELLO, clientLanguage });
export const bonjour = (serverLanguage: string): Message => ({ type: MSG_BONJOUR, serverLanguage });
export const ping = (timestampMs: bigint): Message => ({ type: MSG_PING, timestampMs });
export const pong = (timestampMs: bigint): Message => ({ type: MSG_PONG, timestampMs });
export const timeNotif = (timestampMs: bigint, iso8601: string): Message => ({ type: MSG_TIME_NOTIFICATION, timestampMs, iso8601 });
export const kissRequest = (osName: string, osVersion: string, osRelease: string, osArch: string): Message => ({ type: MSG_KISS_REQUEST, osName, osVersion, osRelease, osArch });
export const kissResponse = (kissLanguage: string, kissEncoding: string, kissTimeZone: string): Message => ({ type: MSG_KISS_RESPONSE, kissLanguage, kissEncoding, kissTimeZone });
export const randomNumber = (randomId: bigint, randomNumber: bigint): Message => ({ type: MSG_RANDOM_NUMBER, randomId, randomNumber });
export const hashResponse = (randomId: bigint, hashHex: string): Message => ({ type: MSG_HASH_RESPONSE, randomId, hashHex });
export const disconnect = (disconnectReason: string): Message => ({ type: MSG_DISCONNECT, disconnectReason });
export const error = (errorCode: number, errorMessage: string): Message => ({ type: MSG_ERROR, errorCode, errorMessage });

File: hello-websocket-python/common/codec.py
#!/usr/bin/env python3
"""Hello WebSocket Protocol Codec - Python implementation.

This module implements the canonical binary protocol defined in PROTOCOL.md.
It provides constants, primitive encoders/decoders, message types, frame codec,
and message dispatch for all 13 message types.
"""

import hashlib
import struct
import time
from dataclasses import dataclass, field
from typing import Optional

# ─── Constants ────────────────────────────────────────────────────────────

PORT = 9898
MAGIC = 0x48
VERSION = 0x01
HEADER_LEN = 8
SERVER_LANG = "PYTHON"
CLIENT_LANG = "PYTHON"

# Message types
MSG_HELLO = 0x01
MSG_BONJOUR = 0x02
MSG_ECHO_REQUEST = 0x03
MSG_ECHO_RESPONSE = 0x04
MSG_KISS_REQUEST = 0x05
MSG_KISS_RESPONSE = 0x06
MSG_PING = 0x07
MSG_PONG = 0x08
MSG_TIME_NOTIFICATION = 0x09
MSG_RANDOM_NUMBER = 0x0A
MSG_HASH_RESPONSE = 0x0B
MSG_DISCONNECT = 0x0C
MSG_ERROR = 0x7F

# Error codes
ERR_DECODE = 0x01
ERR_UNKNOWN_MSG_TYPE = 0x02
ERR_TRUNCATED_PAYLOAD = 0x03
ERR_BAD_MAGIC = 0x04
ERR_BAD_VERSION = 0x05
ERR_SESSION_NOT_FOUND = 0x06
ERR_INTERNAL = 0x07

# Intervals (seconds)
PING_INTERVAL = 1.0
SESSION_TIMEOUT = 60.0
TIME_INTERVAL = 5.0
RANDOM_INTERVAL = 5.0
KISS_INTERVAL = 5.0


# ─── Primitive Encoders ──────────────────────────────────────────────────

class ByteWriter:
    def __init__(self):
        self.buf = bytearray()

    def write_u8(self, v: int):
        self.buf.append(v & 0xFF)

    def write_u16(self, v: int):
        self.buf.extend(struct.pack(">H", v))

    def write_u32(self, v: int):
        self.buf.extend(struct.pack(">I", v))

    def write_i32(self, v: int):
        self.buf.extend(struct.pack(">i", v))

    def write_i64(self, v: int):
        self.buf.extend(struct.pack(">q", v))

    def write_string(self, s: str):
        b = s.encode("utf-8")
        self.write_u32(len(b))
        self.buf.extend(b)

    def write_kv(self, m: dict):
        self.write_u32(len(m))
        for k, v in m.items():
            self.write_string(k)
            self.write_string(v)

    def data(self) -> bytes:
        return bytes(self.buf)


# ─── Primitive Decoders ─────────────────────────────────────────────────

class ByteReader:
    def __init__(self, data: bytes):
        self.data = data
        self.pos = 0

    def remaining(self) -> int:
        return len(self.data) - self.pos

    def require(self, size: int, field: str):
        if size < 0 or size > self.remaining():
            raise ValueError(f"{field} requires {size} bytes, only {self.remaining()} remain")

    def read_u8(self) -> int:
        self.require(1, "u8")
        v = self.data[self.pos]
        self.pos += 1
        return v

    def read_u16(self) -> int:
        self.require(2, "u16")
        v = struct.unpack(">H", self.data[self.pos:self.pos + 2])[0]
        self.pos += 2
        return v

    def read_u32(self) -> int:
        self.require(4, "u32")
        v = struct.unpack(">I", self.data[self.pos:self.pos + 4])[0]
        self.pos += 4
        return v

    def read_i32(self) -> int:
        self.require(4, "i32")
        v = struct.unpack(">i", self.data[self.pos:self.pos + 4])[0]
        self.pos += 4
        return v

    def read_i64(self) -> int:
        self.require(8, "i64")
        v = struct.unpack(">q", self.data[self.pos:self.pos + 8])[0]
        self.pos += 8
        return v

    def read_string(self) -> str:
        ln = self.read_u32()
        self.require(ln, "string")
        s = self.data[self.pos:self.pos + ln].decode("utf-8")
        self.pos += ln
        return s

    def read_kv(self) -> dict:
        count = self.read_u32()
        if count > self.remaining() // 8:
            raise ValueError(f"kv count {count} exceeds remaining payload")
        m = {}
        for _ in range(count):
            k = self.read_string()
            v = self.read_string()
            m[k] = v
        return m


# ─── Frame Codec ─────────────────────────────────────────────────────────

def encode_frame(msg_type: int, payload: bytes) -> bytes:
    header = struct.pack(">BBBB", MAGIC, VERSION, msg_type, 0x00)
    header += struct.pack(">I", len(payload))
    return header + payload


class CodecError(ValueError):
    """A decode failure carrying the precise PROTOCOL.md §7 error code.

    Lets callers (e.g. the server's receive loop) classify a decode failure
    exactly instead of pattern-matching the error message text, which
    silently breaks if the wording ever changes.
    """

    def __init__(self, code: int, message: str):
        super().__init__(message)
        self.code = code


def decode_frame(data: bytes):
    if len(data) < HEADER_LEN:
        raise CodecError(ERR_TRUNCATED_PAYLOAD, f"frame too short: {len(data)} bytes")
    magic = data[0]
    if magic != MAGIC:
        raise CodecError(ERR_BAD_MAGIC, f"bad magic: 0x{magic:02x}")
    version = data[1]
    if version != VERSION:
        raise CodecError(ERR_BAD_VERSION, f"bad version: 0x{version:02x}")
    msg_type = data[2]
    payload_len = struct.unpack(">I", data[4:8])[0]
    if payload_len != len(data) - HEADER_LEN:
        raise CodecError(
            ERR_TRUNCATED_PAYLOAD,
            f"payload length mismatch: declared {payload_len}, available {len(data) - HEADER_LEN}",
        )
    return msg_type, data[HEADER_LEN:HEADER_LEN + payload_len]


# ─── Message Types ───────────────────────────────────────────────────────

@dataclass
class Hello:
    client_language: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_string(self.client_language)
        return encode_frame(MSG_HELLO, w.data())


@dataclass
class Bonjour:
    server_language: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_string(self.server_language)
        return encode_frame(MSG_BONJOUR, w.data())


@dataclass
class EchoRequest:
    id: int = 0
    meta: str = ""
    data: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.id)
        w.write_string(self.meta)
        w.write_string(self.data)
        return encode_frame(MSG_ECHO_REQUEST, w.data())


@dataclass
class EchoResult:
    idx: int = 0
    type: int = 0
    kv: dict = field(default_factory=dict)


@dataclass
class EchoResponse:
    status: int = 200
    results: list = field(default_factory=list)

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i32(self.status)
        w.write_u32(len(self.results))
        for r in self.results:
            w.write_i64(r.idx)
            w.write_u8(r.type)
            w.write_kv(r.kv)
        return encode_frame(MSG_ECHO_RESPONSE, w.data())


@dataclass
class KissRequest:
    os_name: str = ""
    os_version: str = ""
    os_release: str = ""
    os_architecture: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_string(self.os_name)
        w.write_string(self.os_version)
        w.write_string(self.os_release)
        w.write_string(self.os_architecture)
        return encode_frame(MSG_KISS_REQUEST, w.data())


@dataclass
class KissResponse:
    language: str = ""
    encoding: str = ""
    time_zone: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_string(self.language)
        w.write_string(self.encoding)
        w.write_string(self.time_zone)
        return encode_frame(MSG_KISS_RESPONSE, w.data())


@dataclass
class Ping:
    timestamp_ms: int = 0

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.timestamp_ms)
        return encode_frame(MSG_PING, w.data())


@dataclass
class Pong:
    timestamp_ms: int = 0

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.timestamp_ms)
        return encode_frame(MSG_PONG, w.data())


@dataclass
class TimeNotification:
    timestamp_ms: int = 0
    iso8601: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.timestamp_ms)
        w.write_string(self.iso8601)
        return encode_frame(MSG_TIME_NOTIFICATION, w.data())


@dataclass
class RandomNumber:
    id: int = 0
    number: int = 0

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.id)
        w.write_i64(self.number)
        return encode_frame(MSG_RANDOM_NUMBER, w.data())


@dataclass
class HashResponse:
    id: int = 0
    hash_hex: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i64(self.id)
        w.write_string(self.hash_hex)
        return encode_frame(MSG_HASH_RESPONSE, w.data())


@dataclass
class Disconnect:
    reason: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_string(self.reason)
        return encode_frame(MSG_DISCONNECT, w.data())


@dataclass
class ErrorMsg:
    code: int = 0
    message: str = ""

    def encode(self) -> bytes:
        w = ByteWriter()
        w.write_i32(self.code)
        w.write_string(self.message)
        return encode_frame(MSG_ERROR, w.data())


# ─── Message Dispatch ────────────────────────────────────────────────────

class Message:
    def __init__(self, msg_type: int):
        self.type = msg_type
        self.hello: Optional[Hello] = None
        self.bonjour: Optional[Bonjour] = None
        self.echo_req: Optional[EchoRequest] = None
        self.echo_resp: Optional[EchoResponse] = None
        self.kiss_req: Optional[KissRequest] = None
        self.kiss_resp: Optional[KissResponse] = None
        self.ping: Optional[Ping] = None
        self.pong: Optional[Pong] = None
        self.time_notif: Optional[TimeNotification] = None
        self.random: Optional[RandomNumber] = None
        self.hash: Optional[HashResponse] = None
        self.disconnect: Optional[Disconnect] = None
        self.error: Optional[ErrorMsg] = None


def decode_message(data: bytes) -> Message:
    msg_type, payload = decode_frame(data)
    r = ByteReader(payload)
    msg = Message(msg_type)

    if msg_type == MSG_HELLO:
        msg.hello = Hello(client_language=r.read_string())
    elif msg_type == MSG_BONJOUR:
        msg.bonjour = Bonjour(server_language=r.read_string())
    elif msg_type == MSG_ECHO_REQUEST:
        msg.echo_req = EchoRequest(id=r.read_i64(), meta=r.read_string(), data=r.read_string())
    elif msg_type == MSG_ECHO_RESPONSE:
        status = r.read_i32()
        count = r.read_u32()
        if count > r.remaining() // 13:
            raise ValueError(f"result count {count} exceeds remaining payload")
        results = []
        for _ in range(count):
            idx = r.read_i64()
            typ = r.read_u8()
            kv = r.read_kv()
            results.append(EchoResult(idx=idx, type=typ, kv=kv))
        msg.echo_resp = EchoResponse(status=status, results=results)
    elif msg_type == MSG_KISS_REQUEST:
        msg.kiss_req = KissRequest(
            os_name=r.read_string(), os_version=r.read_string(),
            os_release=r.read_string(), os_architecture=r.read_string())
    elif msg_type == MSG_KISS_RESPONSE:
        msg.kiss_resp = KissResponse(
            language=r.read_string(), encoding=r.read_string(), time_zone=r.read_string())
    elif msg_type == MSG_PING:
        msg.ping = Ping(timestamp_ms=r.read_i64())
    elif msg_type == MSG_PONG:
        msg.pong = Pong(timestamp_ms=r.read_i64())
    elif msg_type == MSG_TIME_NOTIFICATION:
        msg.time_notif = TimeNotification(timestamp_ms=r.read_i64(), iso8601=r.read_string())
    elif msg_type == MSG_RANDOM_NUMBER:
        msg.random = RandomNumber(id=r.read_i64(), number=r.read_i64())
    elif msg_type == MSG_HASH_RESPONSE:
        msg.hash = HashResponse(id=r.read_i64(), hash_hex=r.read_string())
    elif msg_type == MSG_DISCONNECT:
        msg.disconnect = Disconnect(reason=r.read_string())
    elif msg_type == MSG_ERROR:
        msg.error = ErrorMsg(code=r.read_i32(), message=r.read_string())
    else:
        raise CodecError(ERR_UNKNOWN_MSG_TYPE, f"unknown message type: 0x{msg_type:02x}")

    return msg


# ─── Utility ─────────────────────────────────────────────────────────────

def now_ms() -> int:
    return int(time.time() * 1000)


def now_iso() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def hash_number(num: int) -> str:
    h = hashlib.sha256(str(num).encode("ascii")).hexdigest()
    return h[:10]


def log(name: str, msg: str):
    ts = time.strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{ts}] [INFO] [{name}] {msg}")
